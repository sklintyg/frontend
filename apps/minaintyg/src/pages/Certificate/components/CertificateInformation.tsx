import { Heading } from '@frontend/components'
import { useFormat } from '../../../hooks/useFormat'
import type { CertificateMetadata } from '../../../schema/certificate.schema'

export function CertificateInformation({ issued, issuer, summary, unit }: CertificateMetadata) {
  const { date } = useFormat()
  return (
    <div className="flex flex-col gap-5 md:flex-row md:gap-10">
      {summary && (
        <div>
          <Heading level={3} size="xs" className="mb-0">
            {summary.label}
          </Heading>
          {summary.value}
        </div>
      )}
      <div>
        <Heading level={3} size="xs" className="mb-0 whitespace-nowrap">
          Intyg utf√§rdat
        </Heading>
        {date(issued)}
      </div>
      <div>
        <Heading level={3} size="xs" className="mb-0">
          Skrivet av
        </Heading>
        {issuer.name}, {unit.name}
      </div>
    </div>
  )
}
