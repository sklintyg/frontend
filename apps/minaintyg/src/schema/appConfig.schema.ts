import { z } from 'zod'

export const appConfigLinkSchema = z.object({
  key: z.string(),
  url: z.string(),
  text: z.string(),
  target: z.string(),
})

export const appConfigSchema = z.object({
  useMinifiedJavascript: z.boolean(),
  version: z.string(),
  buildNumber: z.string(),
  elva77MainUrl: z.string(),
  elva77LoginUrl: z.string(),
  applicationLogoutUrl: z.string(),
  miUserSurveyUrl: z.string(),
  miUserSurveyVersion: z.string(),
  miUserSurveyDateFrom: z.string(),
  miUserSurveyDateTo: z.string(),
  knownRecipients: z.array(z.object({ id: z.string(), name: z.string(), trusted: z.boolean() })),
  links: z.record(appConfigLinkSchema),
  banners: z.array(z.unknown()),
})

export type AppConfig = z.infer<typeof appConfigSchema>
export type AppConfigLink = z.infer<typeof appConfigLinkSchema>
