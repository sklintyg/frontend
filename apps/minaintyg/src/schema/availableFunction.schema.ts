import { z } from 'zod'

export const AvailableFunctionType = z.enum(['SEND_CERTIFICATE'])
export const AvailableFunctionInformationType = z.enum([''])

export const availableFunctionInformationSchema = z.object({
  type: AvailableFunctionInformationType,
  id: z.string(),
})

export const availableFunctionSchema = z.object({
  body: z.string().nullable(),
  description: z.string().nullable(),
  name: z.string().nullable(),
  title: z.string().nullable(),
  type: AvailableFunctionType,
  information: z.array(availableFunctionInformationSchema),
})

export type AvailableFunction = z.infer<typeof availableFunctionSchema>
