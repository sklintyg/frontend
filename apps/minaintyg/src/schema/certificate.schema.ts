import { z } from 'zod'

export const CertificateStatusEnum = z.enum(['NEW', 'REPLACED', 'SENT', 'NOT_SENT'])

export const certificateEventSchema = z.object({
  timestamp: z.string().datetime(),
  certificateId: z.string(),
  description: z.string(),
})

export const certificateSummarySchema = z.object({ label: z.string(), value: z.string() }).optional()

export const certificateTypeSchema = z.object({
  id: z.string(),
  version: z.string(),
  description: z.string().optional(),
})

export const certificateIssuerSchema = z.object({
  name: z.string(),
  phoneNumber: z.string(),
})

export const certificateUnitSchema = z.object({
  id: z.string(),
  name: z.string(),
  address: z.string(),
})

export const certificateMetadataSchema = z.object({
  issuer: certificateIssuerSchema,
  unit: certificateUnitSchema,
  name: z.string(),
  events: z.array(certificateEventSchema),
  statuses: z.array(CertificateStatusEnum),
  id: z.string(),
  issued: z.string().datetime(),
  type: certificateTypeSchema,
  summary: certificateSummarySchema,
})

export const certificateContentSchema = z.object({ heading: z.string(), body: z.string() })

export const certificateSchema = z.object({
  metadata: certificateMetadataSchema,
  content: z.array(certificateContentSchema),
})

export type CertificateContent = z.infer<typeof certificateContentSchema>
export type CertificateEvent = z.infer<typeof certificateEventSchema>
export type CertificateIssuer = z.infer<typeof certificateIssuerSchema>
export type CertificateMetadata = z.infer<typeof certificateMetadataSchema>
export type CertificateStatus = z.infer<typeof CertificateStatusEnum>
export type CertificateSummary = z.infer<typeof certificateSummarySchema>
export type CertificateUnit = z.infer<typeof certificateUnitSchema>
export type Certificate = z.infer<typeof certificateSchema>
