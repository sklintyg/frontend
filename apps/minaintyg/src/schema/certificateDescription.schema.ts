import { z } from 'zod'

export const certificateDescriptionSchema = z.object({
  id: z.string(),
  label: z.string(),
  description: z.string(),
  detailedDescription: z.string(),
  issuerTypeId: z.string(),
  cssPath: z.string(),
  scriptPath: z.string(),
  dependencyDefinitionPath: z.string(),
  defaultRecipient: z.string(),
  deprecated: z.boolean(),
  displayDeprecated: z.boolean(),
})

export type CertificateDescription = z.infer<typeof certificateDescriptionSchema>
