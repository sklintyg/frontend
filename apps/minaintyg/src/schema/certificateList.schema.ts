import { z } from 'zod'

export enum CertificateStatus {
  NEW = 'Nytt',
  REPLACED = 'Ers√§tter intyg',
  SENT = 'Skickat',
  NOT_SENT = 'Ej skickat',
}

export const certificateListSummarySchema = z.object({ label: z.string(), value: z.string() }).optional()

export const certificateListEventSchema = z.object({
  timestamp: z.string().datetime(),
  certificateId: z.string(),
  description: z.string(),
})

export const certificateListIssuerSchema = z.object({ name: z.string() })
export const certificateUnitSchema = z.object({ id: z.string(), name: z.string() })
export const certificateTypeSchema = z.object({
  id: z.string(),
  name: z.string(),
  version: z.string(),
})

export const certificateListItemSchema = z.object({
  summary: certificateListSummarySchema,
  issuer: certificateListIssuerSchema,
  unit: certificateUnitSchema,
  events: z.array(certificateListEventSchema),
  statuses: z.array(z.nativeEnum(CertificateStatus)),
  id: z.string(),
  issued: z.string().datetime(),
  type: certificateTypeSchema,
})

export type CertificateListEvent = z.infer<typeof certificateListEventSchema>
export type CertificateListIssuer = z.infer<typeof certificateListIssuerSchema>
export type CertificateListUnit = z.infer<typeof certificateUnitSchema>
export type CertificateListSummary = z.infer<typeof certificateListSummarySchema>
export type CertificateListItem = z.infer<typeof certificateListItemSchema>
