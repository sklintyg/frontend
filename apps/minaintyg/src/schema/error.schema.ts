import { z } from 'zod'

export const ErrorType = z.enum(['unknown', 'login-failed', 'unavailable', 'logged-out', 'not-found', 'underage'])
export type ErrorTypeEnum = z.infer<typeof ErrorType>

export const ErrorPageActionType = z.enum(['start', 'login', '1177', 'about'])
export type ErrorPageActionTypeEnum = z.infer<typeof ErrorPageActionType>

export const ErrorCode = z.enum([
  'CLIENT_ERROR',
  'LOGIN_FAILED',
  'UNKNOWN_INTERNAL_ERROR',
  'FETCH_ERROR',
  'PARSING_ERROR',
  'TIMEOUT_ERROR',
  'CUSTOM_ERROR',
])

export type ErrorCodeEnum = z.infer<typeof ErrorCode>

export const errorSchema = z.object({
  id: z.string(),
  code: ErrorCode.or(z.number()),
  message: z.string(),
  stackTrace: z.string().nullable().optional(),
})

export type ErrorData = z.infer<typeof errorSchema>
