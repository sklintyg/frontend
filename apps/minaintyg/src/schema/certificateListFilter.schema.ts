import { z } from 'zod'
import { CertificateStatus } from './certificateList.schema'

export const certificateFilterOptionsSchema = z.object({
  total: z.number(),
  statuses: z.array(z.nativeEnum(CertificateStatus)),
  units: z.array(z.string()),
  certificateTypes: z.array(z.string()),
  years: z.array(z.string()),
})

type SelectedOptions<T> = { [X in keyof T]: T[X] extends Array<unknown> ? T[X][number] : T[X] }

export type CertificateFilterOptions = z.infer<typeof certificateFilterOptionsSchema>
export type CertificateSelectedOptions = SelectedOptions<CertificateFilterOptions>
