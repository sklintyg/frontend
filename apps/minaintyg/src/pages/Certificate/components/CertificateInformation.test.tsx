import { fakerFromSchema } from '@frontend/fake'
import { render, screen } from '@testing-library/react'
import { Route, RouterProvider, createMemoryRouter, createRoutesFromElements } from 'react-router-dom'
import { certificateMetadataSchema } from '../../../schema/certificate.schema'
import { CertificateInformation } from './CertificateInformation'

function renderComponent(showEvents = true) {
  const metadata = fakerFromSchema(certificateMetadataSchema)({
    issuer: {
      name: 'Mats Andersson',
    },
    unit: {
      id: 'fac5103a-357b-4037-847e-eadf5866573a',
      name: 'Jansson, Lundqvist Kommanditbolag',
      address: 'Löfgrens Väg 262, 904 38 Båhamn',
      phoneNumber: '3226-91101',
    },
    events: [
      {
        timestamp: '2023-09-26T03:28:09.273Z',
        certificateId: 'commodi',
        description: 'Ersätter ett intyg som inte längre är aktuellt',
      },
    ],
    id: '39a4e3af-6f92-46d8-875f-4ca5ff96589f',
    issued: '2022-11-10T22:56:55.241Z',
    type: {
      id: 'ts-bas',
      name: 'Transportstyrelsens läkarintyg högre körkortsbehörighet',
      version: '1',
    },
    summary: {
      label: 'Avser diagnos',
      value: 'Downs syndrom',
    },
  })
  return render(
    <RouterProvider
      router={createMemoryRouter(
        createRoutesFromElements(<Route path="/" element={<CertificateInformation {...metadata} showEvents={showEvents} />} />)
      )}
    />
  )
}

it('Should render as expected', () => {
  const { container } = renderComponent()
  expect(screen.getByText(/ersätter ett intyg som inte längre är aktuellt/i)).toBeInTheDocument()
  expect(container).toMatchSnapshot()
})

it('Should be possible to hide event information', () => {
  const { container } = renderComponent(false)
  expect(screen.queryByText(/ersätter ett intyg som inte längre är aktuellt/i)).not.toBeInTheDocument()
  expect(container).toMatchSnapshot()
})
