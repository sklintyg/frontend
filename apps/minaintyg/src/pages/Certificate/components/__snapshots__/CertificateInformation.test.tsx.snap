// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should render as expected 1`] = `
<div>
  <div
    class="flex flex-col gap-5 md:flex-row md:gap-10"
  >
    <div>
      <h3
        class="ids-heading-4 mb-0"
      >
        Intyg utf√§rdat
      </h3>
      2022-11-10
    </div>
    <div>
      <h3
        class="ids-heading-4 mb-0"
      >
        Avser diagnos
      </h3>
      Downs syndrom
    </div>
    <div>
      <h3
        class="ids-heading-4 mb-0"
      >
        Skrivet av
      </h3>
      Mats Andersson
    </div>
  </div>
</div>
`;
